File: backend/app/schemas/msr_visit.py
from datetime import datetime
from pydantic import BaseModel
from typing import Optional

class MSRVisit(BaseModel):
    # Visit Identifiers
    visit_id: int
    visit_datetime: datetime
    
    # Payment Information
    coupon_amount: float
    
    # Scanner Information (The MSR/Statehead/Zonalhead)
    msr_id: int
    msr_name: str
    msr_role: str
    
    # Mechanic Information (The Visited Entity)
    mechanic_id: int
    mechanic_name: str
    scheduled_lat: Optional[float] = None
    scheduled_lng: Optional[float] = None
    
    # Actual Scanned Location
    scanned_lat: Optional[float] = None
    scanned_lng: Optional[float] = None
    scanned_address: Optional[str] = None
    
    # Location Verification Result
    location_mismatch: bool

    class Config:
        from_attributes = True
        
class MSRVisitGrouped(BaseModel):
    # Unique identifier for this visit event
    scan_batch_id: int
    visit_datetime: datetime
    
    # Scanner Information
    msr_id: int
    msr_name: str
    msr_role: str
    
    # Mechanic Information
    mechanic_id: int
    mechanic_name: str
    scheduled_lat: Optional[float] = None
    scheduled_lng: Optional[float] = None
    
    # Scanned Location (This will be the same for all coupons in the batch)
    scanned_lat: Optional[float] = None
    scanned_lng: Optional[float] = None
    scanned_address: Optional[str] = None
    
    # Visit Summary
    location_mismatch: bool
    total_coupons: int  # Count of coupons scanned in this batch for this mechanic
    total_amount: float # Sum of coupon_value for all coupons in this batch

    class Config:
        from_attributes = True
        

File: frontend/src/admin/types/msrVisit.ts
export interface MSRVisit {
  visit_id: number;
  visit_datetime: string;
  coupon_amount: number;
  msr_id: number;
  msr_name: string;
  msr_role: string;
  mechanic_id: number;
  mechanic_name: string;
  scheduled_lat?: number;
  scheduled_lng?: number;
  scanned_lat?: number;
  scanned_lng?: number;
  scanned_address?: string;
  location_mismatch: boolean;
}

export interface MSRVisitGrouped {
  scan_batch_id: number;
  visit_datetime: string;
  msr_id: number;
  msr_name: string;
  msr_role: string;
  mechanic_id: number;
  mechanic_name: string;
  scheduled_lat?: number;
  scheduled_lng?: number;
  scanned_lat?: number;
  scanned_lng?: number;
  scanned_address?: string;
  location_mismatch: boolean;
  total_coupons: number;
  total_amount: number;
}

export interface MSRVisitsParams {
  start_date: string;
  end_date: string;
  scanner_id?: number;
  scanner_role?: string;
}

export interface MapPin {
  marker: any;
  id: number;
  lat: number;
  lng: number;
  type: 'scheduled' | 'scanned';
  mismatch: boolean;
  visit: MSRVisitGrouped;
  color: string;
}

export interface MSRVisitsTableProps {
  visits: MSRVisitGrouped[];
  selectedPinId?: number;
  onRowClick?: (visit: MSRVisitGrouped) => void;
  dateRangeColors: string[];
  rowRefs?: React.MutableRefObject<Record<number, HTMLTableRowElement | null>>;
}